packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,238 @@
+
+        if (!chunkGroup) {
+          continue
+        }
+
+        for (const chunk of (chunkGroup as any).chunks || []) {
+          chunk.files?.forEach((file: string) => {
+            entryFiles.add(nodePath.join(compilation.outputOptions.path, file))
+          })
+        }
+      }
+
+      assets[
+        `${isWebpack5 ? '../' : ''}${entrypoint.name}.nft.json`
+      ] = new sources.RawSource(
+        JSON.stringify({
+          version: TRACE_OUTPUT_VERSION,
+          files: [...entryFiles, ...this.entryTraces.get(entrypoint.name)!],
+        })
+      )
+    }
+  }
+
+  apply(compiler: webpack.Compiler) {
+    if (isWebpack5) {
+      compiler.hooks.make.tap('NextJsPagesManifest', (compilation) => {
+        // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+        compilation.hooks.processAssets.tap(
+          {
+            name: 'NextJsPagesManifest',
+            // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+            stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONS,
+          },
+          (assets: any) => {
+            this.createTraceAssets(compilation, assets)
+          }
+        )
+      })
+    } else {
+      compiler.hooks.emit.tap('NextJsPagesManifest', (compilation: any) => {
+        this.createTraceAssets(compilation, compilation.assets)
+      })
+    }
+
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.finishModules.tapAsync(
+        PLUGIN_NAME,
+        async (_stats: any, callback: any) => {
+          try {
+            const depModMap = new Map<string, any>()
+
+            compilation.entries.forEach((entry) => {
+              const name = entry.name || entry.options?.name
+
+              if (name?.startsWith('pages/') && entry.dependencies[0]) {
+                const entryMod = getModuleFromDependency(
+                  compilation,
+                  entry.dependencies[0]
+                )
+
+                if (entryMod.resource) {
+                  this.entryNameMap.set(entryMod.resource, name)
+                  this.entryModMap.set(entryMod.resource, entryMod)
+                }
+              }
+            })
+
+            const readFile = (path: string) => {
+              const mod = depModMap.get(path) || this.entryModMap.get(path)
+
+              // map the transpiled source when available to avoid
+              // parse errors in node-file-trace
+              if (mod?._source?._valueAsBuffer) {
+                return mod._source._valueAsBuffer
+              }
+
+              try {
+                return compilation.inputFileSystem.readFileSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT' || e.code === 'EISDIR') {
+                  return null
+                }
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
```

Comment:
sync fs calls might have a performance influence...

Reply:
It seems `node-file-trace` currently relies [on sync fs calls](https://github.com/vercel/nft/blob/f4904fe926bade80cdc2d95121743db10fc75e6a/src/node-file-trace.ts#L158), ideally we could get these from the webpack cache so that it's not actually hitting the filesystem for most of these calls